src/schema/schema_create.c

```diff
@@ -961,6 +975,102 @@ __create_data_source(
     return (dsrc->create(dsrc, &session->iface, uri, (WT_CONFIG_ARG *)cfg));
 }
 
+/*
+ * __import_entry_cmp --
+ *     Qsort function: sort the import entries array by name.
+ */
+static int WT_CDECL
+__import_entry_cmp(const void *a, const void *b)
+{
+    WT_IMPORT_ENTRY *ae, *be;
+
+    ae = (WT_IMPORT_ENTRY *)a;
+    be = (WT_IMPORT_ENTRY *)b;
+
+    return strcmp(ae->name, be->name);
+}
+
+/*
+ * __schema_parse_wt_export --
+ *     Parse export metadata file and populate array of name/config entries. The array is sorted by
+ *     entry name. Caller is responsible to free any memory allocated for the import list.
+ */
+static int
+__schema_parse_wt_export(
+  WT_SESSION_IMPL *session, const char *export_file, WT_IMPORT_LIST *import_list)
+{
+    WT_DECL_ITEM(key);
+    WT_DECL_ITEM(value);
+    WT_DECL_RET;
+    WT_FSTREAM *fs;
+    char *metacfg;
+    const char *md_key, *md_value;
+    bool exist;
+
+    metacfg = NULL;
+
+    /* Look for a metadata export file: if we find it, load it. */
+    WT_RET(__wt_fs_exist(session, export_file, &exist));
+    if (!exist)
+        return (0);
+    WT_RET(__wt_fopen(session, export_file, 0, WT_STREAM_READ, &fs));
+
+    /* Read line pairs and add them to the import list. */
+    WT_ERR(__wt_scr_alloc(session, 1024, &key));
+    WT_ERR(__wt_scr_alloc(session, 1024, &value));
+    for (;;) {
+        WT_ERR(__wt_getline(session, fs, key));
+        if (key->size == 0)
+            break;
+        WT_ERR(__wt_getline(session, fs, value));
+        if (value->size == 0)
+            WT_ERR_PANIC(session, EINVAL, "%s: zero-length value", export_file);
+
+        /* Skip any WiredTiger tables and system entries. */
+        md_key = (const char *)key->data;
+        md_value = (const char *)value->data;
+        if (WT_SUFFIX_MATCH(md_key, WT_METAFILE) || WT_SUFFIX_MATCH(md_key, WT_HS_FILE) ||
+          WT_PREFIX_MATCH(md_key, WT_SYSTEM_PREFIX))
+            continue;
+
+        /*
+         * Check if the entry already exists in the metadata. Raise an error if we try to import an
+         * existing URI rather than just silently returning.
+         */
+        if ((ret = __wt_metadata_search(session, md_key, &metacfg)) != WT_NOTFOUND) {
+            WT_TRET(EEXIST);
+            goto err;
+        }
+
+        if (metacfg != NULL) {
+            __wt_free(session, metacfg);
+        }
+
+        /* Grow the entries array if needed. */
+        WT_ERR(__wt_realloc_def(session, &import_list->entries_allocated,
+          import_list->entries_next + 1, &import_list->entries));
+
```

Comment:
If I'm reading this correctly, we're reading every single entry and copying it into the array. If someone has 10000 tables and wants to import 3 of them, we will be generating and sorting this table each time.

Using the schema code and knowing we're importing from the file, I think the schema code could call something like this function that then searches for the key it needs. That way the create code for `table:name` would also call in for `colgroup:name` and so on to get all the parts needed. So I think this function should change its focus to be searching for a given key and returning the value/config.

Reply:
My focus was to minimize file I/O. If each `__create_colgroup`, `__create_table`, etc will fetch required information directly from `WiredTiger.export` it can be too costly. For instance, if we have many wtobj files. What we can do to find a middle ground here is we can perform one pass over the file and read in only related entries and ignore others. What do you think @sueloverso ?

Reply:
How about this a3af9c34425a9165d516ee164ad452fa6791fad5 ? This change filters out most of the unnecessary entries. However, if we import `table:test`, `table:test123` will be added to the list too. `test123` will not be used in the actual import but they will occupy extra memory. We can make more sophisticated filter to get precise results.

Reply:
I think filtering out most entries is fine. MongoDB would be using `table:collection-<long string of numbers>` so there would not be a concern for overlapping names. You could get the URI substring and filter any entry containing it (i.e. so `table:collection-1` would search for anything with `collection-1` so it would also find `colgroup:collection-1` and `file:collection-1.wt` or `object:collection-1.000001.wtobj`.

Reply:
@AlexeyAnisimovAU - I just looked at the specific commit you posted and it appears you already do what I suggested above. I think that meets the needs.

Reply:
Great, thank you!